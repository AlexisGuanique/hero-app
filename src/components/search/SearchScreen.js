import { useLocation, useNavigate } from 'react-router-dom'
import queryString from 'query-string'
import { useForm } from '../../hooks/useForm'
import { HeroCard } from '../heroList'
import { getHeroesByName } from '../../helpers'


export const SearchScreen = () => {

  // Para hacer la navegacion
  const navigate = useNavigate();

  // Para obtener la informacion de la ubicacion donde nos encontramos
  const location = useLocation();

  // De esta manera nos evitamos procesar nosotros el query enviamo por el formulario
  const { q = '' } = queryString.parse(location.search)

  const heroes = getHeroesByName(q);


  const showSearch = (q.length === 0);
  const showError = (q.length > 0) && heroes.length === 0


  const { searchText, handleInputChange } = useForm({
    searchText: q
  })

  const onSearchSubmit = (evento) => {
    evento.preventDefault();

    // if( searchText.trim().length <= 1) return;

    // De esta manera navegamos entre paginas y enviamos query params
    navigate(`?q=${searchText}`);

  }



  return (
    <div className="p-5">

      <h1>Search</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />


          <form onSubmit={onSearchSubmit} aria-label='form'>
            <input
              type='text'
              placeholder="Busca tu heroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />

            <button className="btn btn-outline-primary mt-1">
              Buscar
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          {/* {
            // Codigo feo
            (heroes.length >= 1)
              ? <div className="alert alert-primary">Busqueda de heroe</div>
              : ( heroes.length === 0 )
                && <div className="alert alert-danger">No se encontro un heroe con la descripcion <b>{q}</b></div>  
          } */}
          
          <div className="alert alert-primary animate__animated animate__bounce" style={{ display: showSearch ? '' : 'none'}}>
            Busqueda de heroe
          </div>

          <div aria-label='alert-danger' className="alert alert-danger animate__animated animate__flash" style={{ display: showError ? '' : 'none'}}>
            No se encontro un heroe con la descripcion <b>{ q }</b>
          </div>


          {
            heroes.map(hero => (
              <HeroCard key={hero.id} {...hero} />
            ))

          }


        </div>

      </div>
    </div>

  )
}
