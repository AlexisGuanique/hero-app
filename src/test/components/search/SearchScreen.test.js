import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { SearchScreen } from '../../../components/search/SearchScreen'


const mockedUseNavigate = jest.fn();


jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));


describe('Pruebas en el <SearchScreen />', () => {

    beforeEach(() => jest.clearAllMocks());

    // Primera prueba
    test('Debe de mostrar correctamente el componente con los valores por defecto', () => {

        const { container } = render(

            <MemoryRouter>
                <SearchScreen />
            </MemoryRouter>

        )

        expect(container).toMatchSnapshot();

    });


    // Segunda prueba
    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {

        const { container } = render(

            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>

        )

        // Verificamos que el valor del input sea el mismo que le pasamos en el queryString
        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        // Verificamos que sea la imagen correcta
        const img = screen.getByRole('img');
        expect(img.src).toBe('http://localhost/assets/heroes/dc-batman.jpg');

        // screen.debug();


        // Verificamos que el div correcto tenga la propiedad de display: none
        const alert = screen.getByLabelText('alert-danger');
        // console.log(alert);
        expect(alert.style.display).toBe('none');


    });

    // Tercera prueba
    test('Debe de mostrar un error si no se encuentra el heroe (batman123)', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
        )

        const alert = screen.getByLabelText('alert-danger');
        expect(alert.style.display).toBe('');

    });


    // Cuarta prueba
    test('Debe de llamar el navigate a la pantalla nueva', () => {

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>

        )

        const input = screen.getByRole('textbox');
        fireEvent.change( input, { target: {name: 'searchText', value: 'superman'}});

        const form = screen.getByRole('form');
        fireEvent.submit( form );

        expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=superman');

    })


});

// TIP
// Para obtener una etiqyeta en especifica utilizamos el aria-label