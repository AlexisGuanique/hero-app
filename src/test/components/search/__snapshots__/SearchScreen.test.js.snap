// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe de mostrar correctamente el componente con los valores por defecto 1`] = `
<div>
  <div
    class="p-5"
  >
    <h1>
      Search
    </h1>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-5"
      >
        <h4>
          Searching
        </h4>
        <hr />
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="Busca tu heroe"
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary mt-1"
          >
            Buscar
          </button>
        </form>
      </div>
      <div
        class="col-7"
      >
        <h4>
          Resultados
        </h4>
        <hr />
        <div
          class="alert alert-primary animate__animated animate__bounce"
        >
          Busqueda de heroe
        </div>
        <div
          aria-label="alert-danger"
          class="alert alert-danger animate__animated animate__flash"
          style="display: none;"
        >
          No se encontro un heroe con la descripcion 
          <b />
        </div>
      </div>
    </div>
  </div>
</div>
`;
